vimiv/utils/imageheader.py

```diff
@@ -0,0 +1,411 @@
+
+Native QT Support:
+| ---    | ---                       | ---                      |
+| Format | Extension according to QT | Vimiv Supported          |
+| ---    | ---                       | ---                      |
+| BMP    | bmp                       | yes                      |
+| GIF    | gif                       | yes                      |
+| JPG    | jpeg, jpg                 | yes                      |
+| PNG    | png                       | yes                      |
+| PBM    | pbm                       | yes                      |
+| PGM    | pgm                       | yes                      |
+| PPM    | ppm                       | yes                      |
+| XBM    | xbm                       | no (undetectable)        |
+| XPM    | xpm                       | yes                      |
+| SVG    | svg                       | yes                      |
+| SVG    | svgz                      | no (different from svg?) |
+| ---    | ---                       | ---                      |
+
+Extended QT Support:
+| ---    | ---                       | ---                        |
+| Format | Extension according to QT | Vimiv Supported            |
+| ---    | ---                       | ---                        |
+| ICNS   | ico                       | yes                        |
+| ICNS   | icns                      | yes                        |
+| ICNS   | cur                       | no (what are magic bytes?) |
+| JP2    | TODO: ?                   | yes                        |
+| MNG    | TODO: ?                   | no (what are magic bytes?) |
+| TGA    | tga                       | no (undetectable)          |
+| TIFF   | tif, tiff                 | yes                        |
+| WBMP   | wbmp                      | no (what are magic bytes?) |
+| WEBP   | webp                      | yes                        |
+| ---    | ---                       | ---                        |
+
+A great list of magic bytes is provided here:
+https://en.wikipedia.org/wiki/List_of_file_signatures
+"""
+
+import contextlib
+import functools
+
+from typing import Optional, List, Callable, Any, Tuple
+
+from PyQt5.QtGui import QImageReader
+
+from vimiv.utils import log, imagereader
+
+_logger = log.module_logger(__name__)
+
+# List containing all registered check functions
+_registry: List[Tuple[str, Callable[[bytes], bool]]] = []
+
+
+def detect(filename: str) -> Optional[str]:
+    """Determine type of image based on the magic bytes.
+
+    Evaluates each registered check function in the order they were registered. If
+    registered with `priority`, then that check is evaluated before all checks with out
+    `priority`.
+
+    Args:
+        filename: Name of file to determine type of.
+
+    Returns:
+        Filetype or None if unknown.
+    """
+    with open(filename, "rb") as f:
+        header = f.read(32)
+
+    for filetype, check in _registry:
```

Comment:
Flipping the contextlib and the for loop should in principle be possible and avoid a lot of duplicate ignored `IndexErrors`, no? I assume any file smaller than 32 is not a valid image in any case.

Reply:
I thought about this as well. Having it as is, has certainly a negative impact on performance. I do not know if there are no valid images of size less than 32B. But given the many obscure bitmap formats, I would not be surprised if there is😅
Either we leave it as is, or change the checks such that they do not raise any exceptions, or we move the contextlib supression out of the loop and pursue one of the other two solution should ever be an issue encountered. What do you think?

Reply:
It seems like the only three that raise are the `pbm`, `pgm` and `ppm` on an empty file, requiring a total of 3 bytes. Now the only check we have that could require less is `bmp`, unless I read this wrong. So we could even move `bmp` ahead of `p?m` in the checklist, and then be safe moving the contextlib outside?

Reply:
You need to elaborate on this. Why cannot all other check functions raise too?

Reply:
In python, slicing does not raise, you can test this with something simple like:
```python
a = []
print(a[6:12])
```

For the imageheader part, I ran this snippet:
```python
from vimiv.utils import imageheader

path = "notanimage.jpg"

with open(path, "w") as f:
    pass

with open(path, "rb") as f:
    header = f.read(32)

    for filetype, check in imageheader._registry:
        try:
            if check(header, f):
                print("filetype is", filetype)
                break
        except IndexError:
            print(f"{filetype} raises")
```
which gives:
```
pbm raises
pgm raises
ppm raises
```

Reply:
Hmm, I was not aware of that. But also remember that clients can register whatever check function they want 🤔

Reply:
That is a valid point, maybe we should remove the suppress completely then and document that the check functions **may not raise**?

This would then re-writing the p?m checks with slicing, something along
```
    return h[:3] in [b"\x50\x31\x0A", b"\x50\x34\x0A"]
```

Now technically the jpg check could also raise if the file is only exactly the three starting bytes `FF D8 FF`, but I hope this is a technicality that doesn't exist in the wild. (I will regret saying this :joy: )

Reply:
Sorry for the delay...

I tried to tune the checks for performance and made them fail (i.e. return False) fast (as most of them will fail). Because of that, I tried to favour simple comparisons over more complex ones, like `in`. This may lead to less *pythonic* code, which is maybe also less readable. But more performant (hopefully; But Python behaves something weirdly when optimizing for performance).

I have done a bunch of benchmarks comparing many different implementations of the check and the detect functions. I realized that try/except induces zero overhead. And indeed, that was added in python 3.11 ([see 2. point](https://docs.python.org/3.11/whatsnew/3.11.html#misc).

Now I am wondering which approach to take:
- Create check functions that do not raise, by using slices over index access (as you suggested)
- But the checks that raise into a try/except block (I know, it is less beautiful what the contextlib supress thingy, here performance matters more that style 😃).
- Seep it as is, but replace the with contectlib stuff with a try/except block (i.e. use try for each call of the check function separately)

Let me know what you think


The benchmark I used for this:
```python
from vimiv.utils import imageheader
from typing import Optional
import timeit
import contextlib

a = "/tmp/test.bla"

with open(a, "wb") as f:
    f.write(100 * b"\x00")


def detectNative() -> Optional[str]:
    with open(a, "rb") as f:
        header = f.read(32)
        for filetype, check in imageheader._registry:
            if check(header, f):
                return filetype
    return None


def detectSupress() -> Optional[str]:
    with open(a, "rb") as f:
        header = f.read(32)
        for filetype, check in imageheader._registry:
            with contextlib.suppress(IndexError):
                if check(header, f):
                    return filetype
    return None


def detectTry() -> Optional[str]:
    with open(a, "rb") as f:
        header = f.read(32)
        for filetype, check in imageheader._registry:
            try:
                if check(header, f):
                    return filetype
            except IndexError:
                return None
    return None


if __name__ == "__main__":
    NUM_ITER = 10000
    out = timeit.timeit("detectNative()", "from __main__ import detectNative", number=NUM_ITER)
    print(out)
    out = timeit.timeit("detectSupress()", "from __main__ import detectSupress", number=NUM_ITER)
    print(out)
    out = timeit.timeit("detectTry()", "from __main__ import detectTry", number=NUM_ITER)
    print(out)

```

Reply:
This is interesting, I didn't know about the zero-overhead exceptions, although they do of course still induce overhead in case they are caught. However, in your last case, this is essentially like moving `contextlib.suppress` before the for loop. Once caught, we assume the file is too small to be an image, and don't catch multiple exceptions. I fully agree with this.

I am also quite surprised the `contextlib.suppress` adds so much overhead. So I would say we use the `detectTry()` version :blush: 

Thanks for the detailed investigation!!

Reply:
Sure there is overhead when the exception is raised. But that is not much of concern to us, as we expect that no exception is raised 😃

That is a typo. I mean to have a `pass/continue` in the except body. The reasoning is again that a check that raises should not affect other checks. This is important as clients can register whatever check function they like.

I have made these last adoptions. You may review and merge if you are happy about it 😊

Reply:
Ah, yes, I always forget about the custom check functions! Thanks, looks great, I only pushed a few tiny style changes. Will merge once the tests pass :blush: